libs/ml/src/ops/matrix_multiply.cpp

```diff
@@ -0,0 +1,418 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2020 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "math/matrix_operations.hpp"
+#include "math/tensor/tensor.hpp"
+#include "ml/exceptions/exceptions.hpp"
+#include "ml/ops/matrix_multiply.hpp"
+
+#include <cassert>
+
+namespace fetch {
+namespace ml {
+namespace ops {
+
+template <class T>
+MatrixMultiply<T>::MatrixMultiply(const SPType &sp)
+  : Ops<T>(sp)
+{
+  transpose_a_ = sp.transpose_a;
+  transpose_b_ = sp.transpose_b;
+}
+
+template <class T>
```

Comment:
generally I have been using template <typename TensorType> rather than <class T> but this is mostly a matter of style. In this context class vs. typename makes almost no difference (unless we do a template template, but we are not going to do that!!!)

Reply:
Did you mean typename?

Reply:
Ah, I see - was confused with first sentence

Reply:
Done